Sources/Algorithms/Suffix.swift

```diff
@@ -0,0 +1,31 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// Suffix(while:)
+//===----------------------------------------------------------------------===//
+
+extension BidirectionalCollection {
+  @inlinable
+  public __consuming func Suffix(while predicate: (Element) throws -> Bool
+  ) rethrows -> [Element] {
+   var result = ContiguousArray<Element>()
+   self.reverse()
+    for element in self {
+      guard try predicate(element) else {
+        break
+      }
+      result.append(element)
+    }
+    result.reverse()
+    return Array(result)
```

Comment:
This double-reverse operation and converting to `Array` shouldn’t be necessary. The last index of the collection can be retrieved with `endIndex`, then walked backwards using `index(before:)`.

This would be similar to the implementation of `prefix(while:)`<sup>[1](https://github.com/apple/swift/blob/main/stdlib/public/core/Collection.swift#L1330)</sup>, but a little bit trickier. Since `endIndex` is beyond the subscript-able range of a collection, the index has to be moved backwards _before_ using it to subscript the collection, and ensuring it doesn’t go before the first element (start index). In SuffixTests.swift, I made some comments about test cases that will be important for ensuring that these cases are accounted for.

----

1. https://github.com/apple/swift/blob/main/stdlib/public/core/Collection.swift#L1330